README.md

```diff
@@ -87,6 +87,50 @@ The PMEM-CSI driver supports running in different modes, which can be controlled
+down. Data is retained until the volume gets deleted.
+* Ephemeral Volumes
+An application is assigned to a node and when it starts to run, a new volume is created for it on that node. When the application stops, the volume is deleted. This [diagram](/docs/images/sequence/pmem-csi-ephemeral-sequence-diagram.png) illustrates how an ephemeral volume gets provisioned in Kubernetes using PMEM-CSI driver.
+* Cache Volumes
+Volumes are created on a certain set of nodes, each with its own local data. Applications are started on those nodes and then get to use the volume on their node. Data persists across application restarts. This
+is useful when the data is only cached information that can be discarded and reconstructed at any time *and* the application can reuse existing local data when restarting. This [diagram](/docs/images/sequence/pmem-csi-cache-sequence-diagram.png) illustrates how a cache volume gets provisioned in Kubernetes using PMEM-CSI driver.
+
+Volume | Kubernetes | PMEM-CSI | Limitations
+--- | --- | --- | ---
+Persistent | supported | supported | topology aware scheduling
+Ephemeral | [in design](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20190122-csi-inline-volumes.md#proposal) | supported | topology aware scheduling, resource constraints
+Cache | supported | supported | topology aware scheduling
+
+[Topology aware scheduling](https://github.com/kubernetes/enhancements/issues/490)
+ensures that an application runs on a node where the volume was created. This is obviously needed for persistent volumes, but also for
+delayed persistent volumes because otherwise a restarted application
+might end up on the wrong node. Usage of cache volumes is more flexible,
+but an application still can only run on one of the nodes where the
+a cache volume is available. For CSI-based drivers like PMEM-CSI,
+Kubernetes >= 1.13 is needed. On older Kubernetes releases, pods must
+be scheduled manually onto the right node(s).
+
+[Topology aware volume provisioning](https://kubernetes.io/blog/2018/10/11/topology-aware-volume-provisioning-in-kubernetes/)
+is needed for delayed persistent volumes.
+
+The upstream design for ephemeral volumes currently does not take
+[resource constraints](https://github.com/kubernetes/enhancements/pull/716#discussion_r250536632)
+into account. If an application gets scheduled onto a node and then
+creating the ephemeral volume on that node fails, the application on
+the node cannot start until resources become available. The same
+applies to how this is currently implemented in PMEM-CSI: a fake
+volume gets created initially, without actually allocating space. Then
+when an application starts and it is time to mount that "existing"
+volume, an attempt is made to allocate the space required for the
+volume locally. Only if there is enough space can application startup
```

Comment:
Hmm, may be Yes. I was thinking to make it use for illustrating the normal persistent volume type, But may not be needed as Kubernetes users will have that knowledge of how normal  dynamic PVs work.

Reply:
But the current diagram isn't the "normal" PVC, is it? It's the fake PVC with delayed allocation on pod startup.

Do you intend to keep this special mode? I thought we agreed to implement the normal PVC mode (allocate space in `CreateVolume`) with low priority, because it isn't really needed.

Reply:
My intention was that reusing the old diagram by making needed changes according to normal PVC flow. There is no fake/delayed volume creation anymore.

I would request you please review the PR *code* to get how really the cache and ephemeral are implemented.

Reply:
As I said in my other comment, it is easier to review how this is meant to work by reading the user-visible documentation for it.

It makes sense to update the existing diagram. Right now it still shows that `CreateVolume` doesn't actually create a volume. Do we agree after merging this PR, `CreateVolume` will end up creating the pmem device on some node?